File: 剑指 Offer 36. 二叉搜索树与双向链表.py
# BST二叉搜索树的性质, 辅助单调栈

class Solution:
    def treeToDoublyList(self, root: 'Node') -> 'Node':
        if not root:
            return
        stack = [root]
        while stack[-1].left:
            stack.append(stack[-1].left)
        head = pre = stack.pop()
        while True:
            if pre.right:
                stack.append(pre.right)
                while stack[-1].left:
                    stack.append(stack[-1].left)
            if not stack:
                break
            cur = stack.pop()
            pre.right = cur
            cur.left = pre
            pre = cur
        pre.right = head
        head.left = pre
        return head

File: 剑指 Offer 36. 二叉搜索树与双向链表.cpp
// BST二叉搜索树的性质, DFS

class Solution
{
public:
    Node *treeToDoublyList(Node *root)
    {
        if (root == nullptr)
            return nullptr;
        dfs(root);
        head->left = pre;
        pre->right = head;
        return head;
    }

private:
    Node *pre, *head;
    void dfs(Node *cur)
    {
        if (cur == nullptr)
            return;
        dfs(cur->left);

        if (pre != nullptr)
        {
            pre->right = cur;
            cur->left = pre;
        }
        else
            head = cur;
        pre = cur;

        dfs(cur->right);
    }
};
